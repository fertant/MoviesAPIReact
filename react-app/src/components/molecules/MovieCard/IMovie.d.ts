export interface IMovieEditState {
  editOpened: boolean,
  theme: any,
}

export interface IMovie {
  id: number,
  title: string,
  subtitle: string,
  description: string,
  rate?: number,
  duration: number,
  img?: string,
  yearOfRelease: string,
  genre: Array<string>,
}

export interface IMovieProps {
  item: IMovie,
}
