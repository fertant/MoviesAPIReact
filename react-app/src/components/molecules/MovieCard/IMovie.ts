export interface IMovieEditState {
  editOpened: boolean;
  theme: any;
}

export interface IMovie {
  id: number;
  title: string;
  subtitle: string;
  description: string;
  rate?: number;
  duration: number;
  img?: string;
  yearOfRelease: number;
  genre: Array<string>
}

export interface IMovieProps {
  item: IMovie;
}
