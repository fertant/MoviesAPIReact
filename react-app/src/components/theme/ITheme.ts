export interface ITheme {
  theme: any
}
