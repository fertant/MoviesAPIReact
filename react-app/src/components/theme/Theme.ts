const theme = {
  color_a: '#f65261',
  color_b: '#424242',
  color_c: '#232323',
  color_d: '#555555',
  color_e: '#ffffff',
  color_f: '#000000',
  color_g: '#f00',
};

export default theme;
