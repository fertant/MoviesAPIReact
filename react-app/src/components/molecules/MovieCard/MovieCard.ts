import styled from 'styled-components';

import { IMovieEditState } from './IMovie';

export const MovieItemWrapper = styled.div`
  position: relative;
  color: ${props => props.theme.color_d};;
  width: 100%;

  @media (min-width: 768px) {
    width: 45%;
  }

  @media (min-width: 1024px) {
    width: 28%;
  }
`;

export const MovieTitleWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding-top: 10px;
`;

export const MovieTitle = styled.h1`
  font-size: 18px;
  font-weight: 200;
  margin: auto 0;
`;

export const MovieYear = styled.div`
  border: 1px ${props => props.theme.color_d} solid;
  border-radius: 3px;
  padding: 2px 0;
  margin: 5px;
  text-align: center;
  width: 70px;
`;

export const MovieScreen = styled.img`
  width: 100%;
  object-fit: cover;
`;

export const MovieGenre = styled.div`
  font-size: 14px;
`;

export const EditIcon = styled.div`
  position: absolute;
  background-color: ${props => props.theme.color_c};
  border-radius: 18px;
  font-size: 0;
  padding: 18px;
  top: 10px;
  right: 10px;
`;

export const ThreeDots = styled.span`
  position: relative;
  background-color: ${props => props.theme.color_e};
  border-radius: 2px;
  font-size: 0;
  padding: 2px;

  &:before {
    position: absolute;
    left: 0px;
    top: -8px;
    content: '';

    background-color: ${props => props.theme.color_e};
    border-radius: 2px;
    font-size: 0;
    padding: 2px;
  }

  &:after {
    position: absolute;
    left: 0;
    top: 8px;
    content: '';

    background-color: ${props => props.theme.color_e};
    border-radius: 2px;
    font-size: 0;    
    padding: 2px;
  }
`;

export const EditMenuItem = styled.button`
  width: 100%;
  padding: 5px 25px;
  margin: 5px 0;
  background: ${props => props.theme.color_c};
  text-align: start;
  border: none;
  font-size: 18px;
  color: ${props => props.theme.color_e};

  &:hover {
    background-color: ${props => props.theme.color_a};
    color: ${props => props.theme.color_e};
  }
`;

export const EditMenu = styled.div`
  display: ${(props: IMovieEditState) => props.editOpened ? 'inline-block' : 'none'};
  position: absolute;
  width: 210px;
  padding: 30px 0 20px;
  background-color: ${props => props.theme.color_c};
  border-radius: 2px;
  top: 10px;
  right: 10px;
`;

export const CloseIcon = styled.button`
  position: absolute;
  right: 5px;
  top: 5px;
  opacity: 0.3;
  border: none;
  background-color: ${props => props.theme.color_c};

  &:hover {
    opacity: 1;
  }

  &:before, &:after {
    position: absolute;
    top: 0;
    content: ' ';
    height: 15px;
    width: 2px;
    background-color: ${props => props.theme.color_e};
  }

  &:before {
    transform: rotate(45deg);
  }
  &:after {
    transform: rotate(-45deg);
  }
`;  
  
export default MovieItemWrapper;
