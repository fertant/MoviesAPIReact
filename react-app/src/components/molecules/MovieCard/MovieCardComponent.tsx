import React, { FunctionComponent, useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { IMovieProps } from './IMovie';
import {
  MovieItemWrapper,
  MovieScreen,
  MovieTitleWrapper,
  MovieTitle,
  EditIcon,
  CloseIcon,
  EditMenu,
  EditMenuItem,
  ThreeDots,
  MovieYear,
  MovieGenre,
} from './MovieCard';
import { actionControlVisibility, actionSelectMovie } from '../../../actions/Actions';

const MovieCardComponent: FunctionComponent<IMovieProps> = ({ item }) => {
  const [editMenuOpened, setEditActive] = useState(false);
  const dispatch = useDispatch();
  const history = useHistory();

  const onEditMovie = () => {
    dispatch(actionSelectMovie(item));
    dispatch(actionControlVisibility('edit', true));
    setEditActive(false);
  };
  const onDeleteMovie = () => {
    dispatch(actionSelectMovie(item));
    dispatch(actionControlVisibility('delete', true));
    setEditActive(false);
  };
  const onShowMovieDetails = useCallback(() => {
    dispatch(actionSelectMovie(item));
    dispatch(actionControlVisibility('details', true));
    history.push(`/film/${item.id}`);
  }, [item]);

  return (
    <MovieItemWrapper>
      <EditIcon onClick={() => setEditActive(true)}>
        <ThreeDots />
      </EditIcon>
      <EditMenu editOpened={editMenuOpened}>
        <CloseIcon onClick={() => setEditActive(false)} />
        <EditMenuItem onClick={onEditMovie}>Edit</EditMenuItem>
        <EditMenuItem onClick={onDeleteMovie}>Delete</EditMenuItem>
      </EditMenu>
      <MovieScreen src={item.img} onClick={onShowMovieDetails} />
      <MovieTitleWrapper>
        <MovieTitle className="title">{ item.title }</MovieTitle>
        <MovieYear>{ item.yearOfRelease }</MovieYear>
      </MovieTitleWrapper>
      <MovieGenre className="genre">{ item.genre.join(', ') }</MovieGenre>
    </MovieItemWrapper>
  );
};

export default MovieCardComponent;
