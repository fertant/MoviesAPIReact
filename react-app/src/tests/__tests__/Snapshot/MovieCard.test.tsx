import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';

import MovieCardComponent from '../../../components/molecules/MovieCard/MovieCardComponent';
import { IMovie } from '../../../components/molecules/MovieCard/IMovie';
import theme from '../../../components/theme/Theme';
import configureStore from '../../../store/configureStore';

const movieObj: IMovie = {
  id: 1,
  title: 'Moana',
  subtitle: 'Moana',
  description: 'description',
  rate: 5,
  duration: 30,
  yearOfRelease: '2020',
  genre: ['Horror'],
};

test('Test render Movie Card', () => {
  const store = configureStore();
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <MovieCardComponent item={movieObj} />
      </Provider>
    </ThemeProvider>,
  );
  const card = component.toJSON();
  expect(card).toMatchSnapshot();
});
