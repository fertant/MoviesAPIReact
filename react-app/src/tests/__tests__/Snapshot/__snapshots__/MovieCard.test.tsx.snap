// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render Movie Card 1`] = `
<div
  className="sc-bdfBwQ cOTdta"
>
  <div
    className="sc-gKsewC bYukru"
    onClick={[Function]}
  >
    <span
      className="sc-iBPRYJ jdGjQX"
    />
  </div>
  <div
    className="sc-pFZIQ dREdHp"
  >
    <button
      className="sc-jrAGrp mvorh"
      onClick={[Function]}
    />
    <button
      className="sc-fubCfw bDxerx"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="sc-fubCfw bDxerx"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
  <img
    className="sc-eCssSg dgFBAc"
    onClick={[Function]}
  />
  <div
    className="sc-gsTCUz jFfEYb"
  >
    <h1
      className="sc-dlfnbm flxOHZ"
    >
      Moana
    </h1>
    <div
      className="sc-hKgILt jgANG"
    >
      2020
    </div>
  </div>
  <div
    className="sc-jSgupP bCPCUk"
  >
    Horror
  </div>
</div>
`;
